sql/core/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/PartitionPruning.scala

```diff
@@ -147,10 +151,15 @@ object PartitionPruning extends Rule[LogicalPlan] with PredicateHelper {
       case _ => fallbackRatio
     }
 
+    val estimatePruningSideSize = filterRatio * partPlan.stats.sizeInBytes.toFloat
     // the pruning overhead is the total size in bytes of all scan relations
     val overhead = otherPlan.collectLeaves().map(_.stats.sizeInBytes).sum.toFloat
-
-    filterRatio * partPlan.stats.sizeInBytes.toFloat > overhead.toFloat
+    if (canBuildBroadcast) {
+      estimatePruningSideSize > overhead
+    } else {
+      // We need to make sure that otherPlan can be broadcast by size to avoid driver OOM
+      canBroadcastBySize(otherPlan, conf) && estimatePruningSideSize * 0.04 > overhead
```

Comment:
We should explain the rationale clearly. The problem here is we can't reuse the broadcast (because the join type doesn't support broadcast), and doing DPP means running an extra query that may have significant overhead. We need to make the pruning side is very big so that DPP is still worthy.

BTW how do you come up with 0.04?

Reply:
(pruning side stats size) * 0.04 * 0.5(spark.sql.optimizer.dynamicPartitionPruning.fallbackFilterRatio) > 10M(spark.sql.autoBroadcastJoinThreshold)

The (pruning side stats size) around 500M.

Reply:
can we make `0.04` configurable?